Hoarder does not support projects with non-uniqe names

If you apply hoarder to [monix]() and run stash it fill fails with

```
[error] (coreJVM/*:stash) java.lang.AssertionError: assertion failed: Cache already exists in /home/krzysztof/.sbt/0.13/sbt-stash/monix/HEAD/2.11/monix/compile!
[error] (executionJVM/*:stash) java.lang.AssertionError: assertion failed: Cache already exists in /home/krzysztof/.sbt/0.13/sbt-stash/monix/HEAD/2.11/monix-execution/compile!
[error] (catsJVM/*:stash) java.lang.AssertionError: assertion failed: Cache already exists in /home/krzysztof/.sbt/0.13/sbt-stash/monix/HEAD/2.11/monix-cats/compile!
[error] (scalaz72JVM/*:stash) java.lang.AssertionError: assertion failed: Cache already exists in /home/krzysztof/.sbt/0.13/sbt-stash/monix/HEAD/2.11/monix-scalaz-72/compile!
[error] (typesJVM/*:stash) java.lang.AssertionError: assertion failed: Cache already exists in /home/krzysztof/.sbt/0.13/sbt-stash/monix/HEAD/2.11/monix-types/compile!
```

Reported by @vpavkin

--- a/0.13/hoarder/src/main/scala/org/romanowski/HoarderPlugin.scala
+++ b/0.13/hoarder/src/main/scala/org/romanowski/HoarderPlugin.scala
@@ -21,97 +21,109 @@ import sbt._
 
 
 object HoarderKeys extends StashKeys with CachedCiKeys with CachedReleaseKeys {
-  private[romanowski] def internalTask[T: Manifest](name: String) =
-    TaskKey[T](s"hoarder:internal:$name", s"Internal hoarder task: $name. Please do not use.")
+	private[romanowski] def internalTask[T: Manifest](name: String) =
+		TaskKey[T](s"hoarder:internal:$name", s"Internal hoarder task: $name. Please do not use.")
 
 
-  case class CacheSetup(sourceRoots: Seq[File],
-                        classpath: Classpath,
-                        classesRoot: Path,
-                        projectRoot: Path,
-                        analysisFile: File,
-                        relativeCacheLocation: Path,
-                        overrideExistingCache: Boolean,
-                        cleanOutputMode: CleanOutputMode,
-                        zipAnalysisFile: Boolean,
-                        configuration: Configuration,
-                        name: String
-                       ) {
-    def cacheLocation(root: Path) = root.resolve(relativeCacheLocation)
-  }
+	case class CacheSetup(sourceRoots: Seq[File],
+	                      classpath: Classpath,
+	                      classesRoot: Path,
+	                      projectRoot: Path,
+	                      analysisFile: File,
+	                      relativeCacheLocation: Path,
+	                      overrideExistingCache: Boolean,
+	                      cleanOutputMode: CleanOutputMode,
+	                      zipAnalysisFile: Boolean,
+	                      configuration: Configuration,
+	                      name: String
+	                     ) {
+		def cacheLocation(root: Path) = root.resolve(relativeCacheLocation)
+	}
 
-  case class ExportCacheSetup(cacheSetup: CacheSetup, compilationResult: CompilationResult)
+	case class ExportCacheSetup(cacheSetup: CacheSetup, compilationResult: CompilationResult)
 
-  case class ExportedCache(analysis: Path, binaries: Option[Path])
+	case class ExportedCache(analysis: Path, binaries: Option[Path])
 
-  val cleanOutputMode = SettingKey[CleanOutputMode]("hoarder:cleanOutputMode", "What should be cleaned prior to cache extraction")
-  val zipAnalysisCache = SettingKey[Boolean]("hoarder:zipAnalysisFile", "Determines if analysis file will be zipped or not")
-  val overrideExistingCache = SettingKey[Boolean]("hoarder:overrideExistingCache", "Override existing stash")
-  val enabledConfigurations = SettingKey[Seq[Configuration]]("hoarder:enabledConfigurations",
-    "Configuration that hoarder will use")
+	val cleanOutputMode = SettingKey[CleanOutputMode]("hoarder:cleanOutputMode", "What should be cleaned prior to cache extraction")
+	val zipAnalysisCache = SettingKey[Boolean]("hoarder:zipAnalysisFile", "Determines if analysis file will be zipped or not")
+	val overrideExistingCache = SettingKey[Boolean]("hoarder:overrideExistingCache", "Override existing stash")
+	val enabledConfigurations = SettingKey[Seq[Configuration]]("hoarder:enabledConfigurations",
+		"Configuration that hoarder will use")
 
 
-  def withConfiguration(config: Configuration): Seq[Setting[_]] = HoarderPlugin.includeConfiguration(config)
+	def withConfiguration(config: Configuration): Seq[Setting[_]] = HoarderPlugin.includeConfiguration(config)
 
-  private[romanowski] val importCacheSetups = internalTask[Seq[CacheSetup]]("importCacheSetups")
-  private[romanowski] val exportCacheSetups = internalTask[Seq[ExportCacheSetup]]("exportCacheSetups")
-  private[romanowski] val perConfigurationSetup = internalTask[CacheSetup]("perConfigurationSetup")
-  private[romanowski] val perConfigurationExportSetup = internalTask[ExportCacheSetup]("perConfigurationExportSetup")
+	private[romanowski] val importCacheSetups = internalTask[Seq[CacheSetup]]("importCacheSetups")
+	private[romanowski] val exportCacheSetups = internalTask[Seq[ExportCacheSetup]]("exportCacheSetups")
+	private[romanowski] val perConfigurationSetup = internalTask[CacheSetup]("perConfigurationSetup")
+	private[romanowski] val perConfigurationExportSetup = internalTask[ExportCacheSetup]("perConfigurationExportSetup")
 }
 
 object HoarderPlugin extends AutoPlugin {
 
-  override def projectSettings = defaultPerProject ++ Stash.settings
+	override def projectSettings = defaultPerProject ++ Stash.settings
 
-  override def globalSettings: Seq[_root_.sbt.Def.Setting[_]] = Stash.globalSettings ++ defaultsGlobal
+	override def globalSettings: Seq[_root_.sbt.Def.Setting[_]] = Stash.globalSettings ++ defaultsGlobal
 
-  override def trigger: PluginTrigger = AllRequirements
+	override def trigger: PluginTrigger = AllRequirements
 
-  object autoImport {
-    val hoarder = HoarderKeys
-  }
+	object autoImport {
+		val hoarder = HoarderKeys
+	}
 
-  import HoarderKeys._
-
-
-  private[romanowski] def includeConfiguration(config: Configuration): Seq[Setting[_]] = {
-    inConfig(config)(Seq(
-      perConfigurationSetup := projectSetupFor.value,
-      perConfigurationExportSetup := ExportCacheSetup(perConfigurationSetup.value, compileIncremental.value)
-    )) ++ Seq(
-      importCacheSetups += perConfigurationSetup.in(config).value,
-      exportCacheSetups += perConfigurationExportSetup.in(config).value,
-      enabledConfigurations += config
-    )
-  }
-
-  private def projectSetupFor = Def.task[CacheSetup] {
-    CacheSetup(
-      sourceRoots = managedSourceDirectories.value ++ unmanagedSourceDirectories.value,
-      classpath = externalDependencyClasspath.value,
-      classesRoot = classDirectory.value.toPath,
-      projectRoot = baseDirectory.value.toPath,
-      analysisFile = (streams in compileIncSetup).value.cacheDirectory / compileAnalysisFilename.value,
-      relativeCacheLocation = Paths.get(name.value).resolve(configuration.value.name),
-      overrideExistingCache = overrideExistingCache.value,
-      cleanOutputMode = cleanOutputMode.value,
-      zipAnalysisFile = zipAnalysisCache.value,
-      configuration = configuration.value,
-      name = name.value
-    )
-  }
+	import HoarderKeys._
 
 
-  def defaultsGlobal = Seq(
-    cleanOutputMode := CleanClasses,
-    zipAnalysisCache := true,
-    overrideExistingCache := false,
-    importCacheSetups := Nil,
-    exportCacheSetups := Nil
-  )
+	private[romanowski] def includeConfiguration(config: Configuration): Seq[Setting[_]] = {
+		inConfig(config)(Seq(
+			perConfigurationSetup := projectSetupFor.value,
+			perConfigurationExportSetup := ExportCacheSetup(perConfigurationSetup.value, compileIncremental.value)
+		)) ++ Seq(
+			importCacheSetups += perConfigurationSetup.in(config).value,
+			exportCacheSetups += perConfigurationExportSetup.in(config).value,
+			enabledConfigurations += config
+		)
+	}
 
-  def defaultPerProject =
-    Seq(importCacheSetups := Nil, exportCacheSetups := Nil, enabledConfigurations := Nil) ++
-      includeConfiguration(Compile) ++
-      includeConfiguration(Test)
+	private def projectSetupFor = Def.task[CacheSetup] {
+		val relativeCacheLocation = {
+			val rootDir = Paths.get(".").toAbsolutePath.getParent
+			val baseDir = baseDirectory.value.toPath.toAbsolutePath
+			val relativePath =
+				if (rootDir == baseDir) Paths.get(name.value)
+				else {
+					assert(baseDir.startsWith(rootDir))
+					rootDir.relativize(baseDir)
+				}
+			relativePath.resolve(configuration.value.name)
+		}
+
+		CacheSetup(
+			sourceRoots = managedSourceDirectories.value ++ unmanagedSourceDirectories.value,
+			classpath = externalDependencyClasspath.value,
+			classesRoot = classDirectory.value.toPath,
+			projectRoot = baseDirectory.value.toPath,
+			analysisFile = (streams in compileIncSetup).value.cacheDirectory / compileAnalysisFilename.value,
+			relativeCacheLocation = relativeCacheLocation,
+			overrideExistingCache = overrideExistingCache.value,
+			cleanOutputMode = cleanOutputMode.value,
+			zipAnalysisFile = zipAnalysisCache.value,
+			configuration = configuration.value,
+			name = name.value
+		)
+	}
+
+
+	def defaultsGlobal = Seq(
+		cleanOutputMode := CleanClasses,
+		zipAnalysisCache := true,
+		overrideExistingCache := false,
+		importCacheSetups := Nil,
+		exportCacheSetups := Nil
+	)
+
+	def defaultPerProject =
+		Seq(importCacheSetups := Nil, exportCacheSetups := Nil, enabledConfigurations := Nil) ++
+			includeConfiguration(Compile) ++
+			includeConfiguration(Test)
 }

--- a/0.13/hoarder/src/main/scala/org/romanowski/hoarder/actions/Stash.scala
+++ b/0.13/hoarder/src/main/scala/org/romanowski/hoarder/actions/Stash.scala
@@ -26,6 +26,7 @@ object Stash extends HoarderEngine {
 			stashImpl,
 			stashApplyImpl,
 			stashCleanImpl,
+			aggregate.in(stashCleanKey) := false,
 			aggregate.in(stashKey) := true,
 			aggregate.in(stashApplyKey) := true
 		)

--- a/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/build.sbt
+++ b/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/build.sbt
@@ -4,4 +4,8 @@ lazy val baseProject = project
 lazy val finalProject = project dependsOn(leafProject)
 lazy val leafProject = project  dependsOn(baseProject)
 
+lazy val nested = project.in(file("js") / "baseProject").settings(
+	name := "baseProject"
+)
+
 val root = project aggregate(leafProject, baseProject, finalProject)

--- a/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/js/baseProject/build.sbt
+++ b/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/js/baseProject/build.sbt
@@ -1,0 +1,1 @@
+libraryDependencies += "com.typesafe.akka" %% "akka-actor" % "2.4.14"

--- a/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/js/baseProject/src/main/scala/base/Base.scala
+++ b/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/js/baseProject/src/main/scala/base/Base.scala
@@ -1,0 +1,20 @@
+/*
+ * Hoarder - Cached compilation plugin for sbt.
+ * Copyright 2016 - 2017, Krzysztof Romanowski
+ * This software is released under the terms written in LICENSE.
+ */
+
+package base
+
+import akka.actor.Actor
+
+
+class Base {
+
+}
+
+class OldActor extends Actor {
+  override def receive = {
+    case _ => println("Ala")
+  }
+}

--- a/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/js/baseProject/src/test/scala/base/BaseTest.scala
+++ b/0.13/hoarderTests/src/sbt-test/stash/multiProjectTest/js/baseProject/src/test/scala/base/BaseTest.scala
@@ -1,0 +1,11 @@
+/*
+ * Hoarder - Cached compilation plugin for sbt.
+ * Copyright 2016 - 2017, Krzysztof Romanowski
+ * This software is released under the terms written in LICENSE.
+ */
+
+package base
+
+object BaseTest {
+  val b: Base = ???
+}

